File: 1760-minimum-limit-of-balls-in-a-bag/1760-minimum-limit-of-balls-in-a-bag.java
class Solution {
    public int minimumSize(int[] nums, int maxOperations) {
        int maxNum = 0;
        for(int num : nums)
            maxNum = Math.max(maxNum, num);
        
        int lo = 1, hi = maxNum;
        while(lo < hi){
            int mid = lo + (hi - lo) / 2;
            
            int ops = 0;
            for(int num : nums)
                ops += (num - 1) / mid;
            
            if(ops > maxOperations)
                lo = mid + 1;
            else
                hi = mid;
        }
        return lo;
    }
}

File: 1760-minimum-limit-of-balls-in-a-bag/1760-minimum-limit-of-balls-in-a-bag.py
class Solution:
    def minimumSize(self, nums: List[int], maxOperations: int) -> int:
        def valid(guess: int) -> bool:
            ops = 0
            for num in nums:
                ops += (num - 1) // guess
                if ops > maxOperations:
                    return False
            return True

        lo, hi = 0, max(nums)
        while lo + 1 < hi:
            mid = lo + (hi - lo) // 2

            if valid(mid):
                hi = mid
            else:
                lo = mid
        return hi